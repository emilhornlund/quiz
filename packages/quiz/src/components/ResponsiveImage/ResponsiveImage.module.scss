@use '../../styles/colors';
@use '../../styles/helpers';
@use '../../styles/variables';

.container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  max-width: 100%;
  max-height: 100%;

  img {
    display: flex;
    max-width: 100%;
    min-height: 100%;
    box-sizing: border-box;
    border: helpers.pxToRem(4px) solid colors.$white-1-50;
    background-color: colors.$white-1-50;
    object-fit: contain;

    @include helpers.mobile {
      border-radius: variables.$mobile-narrow-layout-gutter;
    }

    @include helpers.tablet {
      border-radius: variables.$tablet-narrow-layout-gutter;
    }

    @include helpers.desktop {
      border-radius: variables.$desktop-narrow-layout-gutter;
    }
  }
}