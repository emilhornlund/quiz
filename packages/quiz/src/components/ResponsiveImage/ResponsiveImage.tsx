import { faTriangleExclamation } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import React, {
  FC,
  ReactNode,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react'
import { useDebounceCallback, useResizeObserver } from 'usehooks-ts'

import { LoadingSpinner, Typography } from '../index.ts'

import styles from './ResponsiveImage.module.scss'

export interface ResponsiveImageProps {
  imageURL?: string
  alt?: string
  noBorder?: boolean
  children?: ReactNode | ReactNode[]
}

type Phase = 'idle' | 'loading' | 'ready' | 'error'
type Size = { width?: number; height?: number }

const ResponsiveImage: FC<ResponsiveImageProps> = ({
  imageURL,
  alt,
  noBorder = false,
  children,
}) => {
  const ref = useRef<HTMLDivElement>(null)

  const [containerSize, setContainerSize] = useState<Size>({
    width: 0,
    height: 0,
  })
  const [intrinsic, setIntrinsic] = useState({ w: 0, h: 0 })
  const [phase, setPhase] = useState<Phase>('idle')
  const [displaySrc, setDisplaySrc] = useState<string | null>(null)

  const applyContainerSize = useCallback((next: Size) => {
    setContainerSize((prev) => {
      const w = Math.round(next.width ?? 0)
      const h = Math.round(next.height ?? 0)
      if ((prev.width ?? 0) === w && (prev.height ?? 0) === h) return prev
      return { width: w, height: h }
    })
  }, [])

  const onResize = useDebounceCallback(applyContainerSize, 80)

  useResizeObserver<HTMLDivElement>({
    // @ts-expect-error (usehooks-ts React 19 types mismatch)
    ref,
    onResize,
    box: 'content-box',
  })

  useEffect(() => {
    if (!imageURL) {
      setPhase('idle')
      setDisplaySrc(null)
      setIntrinsic({ w: 0, h: 0 })
      return
    }

    setPhase('loading')
    setDisplaySrc(null)
    setIntrinsic({ w: 0, h: 0 })

    const img = new Image()
    img.onload = () => {
      setIntrinsic({ w: img.naturalWidth || 0, h: img.naturalHeight || 0 })
      setDisplaySrc(imageURL)
      setPhase('ready')
    }
    img.onerror = () => {
      setDisplaySrc(null)
      setIntrinsic({ w: 0, h: 0 })
      setPhase('error')
    }
    img.src = imageURL

    return () => {
      img.onload = null
      img.onerror = null
    }
  }, [imageURL])

  const box = useMemo(() => {
    if (phase !== 'ready' || !intrinsic.w || !intrinsic.h) return null

    const cw = containerSize.width ?? 0
    const ch = containerSize.height ?? 0
    if (cw <= 0 && ch <= 0) return null

    const { w: iw, h: ih } = intrinsic
    const rw = cw > 0 ? cw / iw : Number.POSITIVE_INFINITY
    const rh = ch > 0 ? ch / ih : Number.POSITIVE_INFINITY
    const s = Math.min(rw, rh)

    if (!Number.isFinite(s) || s <= 0) return null
    return { w: iw * s, h: ih * s }
  }, [phase, containerSize, intrinsic])

  return (
    <div ref={ref} className={styles.container}>
      {phase === 'ready' && displaySrc && box && (
        <div
          className={noBorder ? styles.boxNoBorder : styles.box}
          style={{ width: box.w, height: box.h }}>
          <img src={displaySrc} alt={alt ?? ''} className={styles.img} />
          {children && <div className={styles.overlay}>{children}</div>}
        </div>
      )}

      {phase === 'loading' && (
        <div className={styles.centerOverlay}>
          <LoadingSpinner />
        </div>
      )}

      {phase === 'error' && (
        <div className={styles.centerOverlay}>
          <div className={styles.inner}>
            <FontAwesomeIcon
              icon={faTriangleExclamation}
              className={styles.icon}
            />
            <Typography variant="subtitle">
              Oh no! Unable to load image
            </Typography>
          </div>
        </div>
      )}
    </div>
  )
}

export default ResponsiveImage
