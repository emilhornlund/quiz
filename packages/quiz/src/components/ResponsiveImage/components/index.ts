export { ImageSquareEffect } from './ImageSquareEffect.tsx'
