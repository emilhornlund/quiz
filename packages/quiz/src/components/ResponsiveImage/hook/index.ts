import { useImageBlurEffect } from './useImageBlurEffect.ts'
export { useImageBlurEffect }
