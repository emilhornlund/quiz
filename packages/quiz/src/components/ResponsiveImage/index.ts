export type { ResponsiveImageProps } from './ResponsiveImage'
export { default } from './ResponsiveImage'
