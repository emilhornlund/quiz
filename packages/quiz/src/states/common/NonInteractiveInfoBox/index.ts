export type { NonInteractiveInfoBoxProps } from './NonInteractiveInfoBox'
export { default } from './NonInteractiveInfoBox'
