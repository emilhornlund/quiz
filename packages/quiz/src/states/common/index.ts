export type { NonInteractiveInfoBoxProps } from './NonInteractiveInfoBox'
export type { PointBehindIndicatorProps } from './PointsBehindIndicator'
export type { QuestionAnswerPickerProps } from './QuestionAnswerPicker'
export type { QuestionMediaProps } from './QuestionMedia'
export type { QuestionTextPreviewProps } from './QuestionTextPreview'
export type { QuestionTypePointsBarProps } from './QuestionTypePointsBar'

export { default as GamePage } from './GamePage'
export { default as NonInteractiveInfoBox } from './NonInteractiveInfoBox'
export { default as PointsBehindIndicator } from './PointsBehindIndicator'
export { default as QuestionAnswerPicker } from './QuestionAnswerPicker'
export { default as QuestionMedia } from './QuestionMedia'
export { default as QuestionTextPreview } from './QuestionTextPreview'
export { default as QuestionTypePointsBar } from './QuestionTypePointsBar'
