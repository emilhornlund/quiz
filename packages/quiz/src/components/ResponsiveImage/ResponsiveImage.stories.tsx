import type { Meta, StoryObj } from '@storybook/react'
import React from 'react'

import ResponsiveImage from './ResponsiveImage'

const meta = {
  component: ResponsiveImage,
  parameters: { layout: 'fullscreen' },
  decorators: [
    (Story) => (
      <div style={{ height: '100vh', display: 'flex', minHeight: 0 }}>
        <div style={{ flex: 1, minHeight: 0 }}>
          <Story />
        </div>
      </div>
    ),
  ],
} satisfies Meta<typeof ResponsiveImage>

export default meta
type Story = StoryObj<typeof meta>

export const Landscape = {
  args: {
    imageURL:
      'https://i.pinimg.com/originals/a6/60/72/a66072b0e88258f2898a76c3f3c01041.jpg',
    alt: 'Who painted The Starry Night?',
  },
} satisfies Story

export const Square = {
  args: {
    imageURL: 'https://wallpapercave.com/wp/wp2824407.jpg',
    alt: 'Who painted The Starry Night?',
  },
} satisfies Story

export const Portrait = {
  args: {
    imageURL: 'https://wallpaperaccess.com/full/157316.jpg',
    alt: 'Who painted The Starry Night?',
  },
} satisfies Story

export const imageEffect = {
  args: {
    imageURL:
      'https://i.pinimg.com/originals/a6/60/72/a66072b0e88258f2898a76c3f3c01041.jpg',
    alt: 'Who painted The Starry Night?',
    effect: 'blur',
    numberOfSquares: 4,
    countdown: {
      initiatedTime: new Date().toISOString(),
      expiryTime: new Date(Date.now() + 30 * 1000).toISOString(),
      serverTime: new Date().toISOString(),
    },
  },
} satisfies Story

export const Error = {
  args: {
    imageURL: 'https://example.com/image.png',
    alt: 'Error loading image',
  },
} satisfies Story
