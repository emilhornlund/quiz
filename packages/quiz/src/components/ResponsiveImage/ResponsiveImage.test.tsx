import { QuestionImageRevealEffectType } from '@quiz/common'
import { act, render, screen } from '@testing-library/react'
import React from 'react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import ResponsiveImage from './ResponsiveImage'

// Image mock: synchronous load/error so we don't need timers.
class MockImage {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onload: null | ((ev?: any) => any) = null
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onerror: null | ((ev?: any) => any) = null
  naturalWidth = 400
  naturalHeight = 300
  set src(val: string) {
    // Treat "bad" and any "invalid-*" as error to match tests using "invalid-url"
    if (!val || val === 'bad' || val.startsWith('invalid')) {
      this.onerror?.(new Event('error'))
    } else {
      this.onload?.(new Event('load'))
    }
  }
}
vi.stubGlobal('Image', MockImage as unknown as typeof Image)

// Fire onResize only once per test to avoid render loops.
let didNotifyResize = false

// useResizeObserver: your component passes onResize that expects {width,height}.
const resizeObserverImpl = vi.fn(
  ({
    onResize,
  }: {
    onResize: (size: { width: number; height: number }) => void
  }) => {
    if (!didNotifyResize) {
      didNotifyResize = true
      onResize({ width: 800, height: 600 })
    }
  },
)

// Debounce: return the original fn (immediate pass-through).
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const debounceImpl = (fn: (...args: any[]) => void) => fn

vi.mock('usehooks-ts', () => ({
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  useResizeObserver: (args: any) => resizeObserverImpl(args),
  useDebounceCallback: (fn: unknown) =>
    debounceImpl(fn as (...args: unknown[]) => void),
}))

// Blur style hook: return a visible style so we can assert.
vi.mock('./hook', () => ({
  useImageBlurEffect: (countdown?: unknown) =>
    countdown ? { filter: 'blur(8px)' } : { filter: 'blur(0px)' },
}))

// Lightweight UI mocks
vi.mock('../index.ts', () => ({
  LoadingSpinner: () => <div data-testid="spinner" />,
  Typography: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="typography">{children}</div>
  ),
}))

// Square effect mock to assert it renders and receives box/effect.
vi.mock('./components', () => ({
  ImageSquareEffect: ({
    box,
    effect,
  }: {
    box: { w: number; h: number }
    effect: QuestionImageRevealEffectType
  }) => (
    <div
      data-testid="square-effect"
      data-w={box.w}
      data-h={box.h}
      data-effect={String(effect)}
    />
  ),
}))

describe('ResponsiveImage', () => {
  beforeEach(() => {
    vi.clearAllMocks()
    didNotifyResize = false
  })

  it('should render ResponsiveImage with props', async () => {
    const { container } = render(
      <ResponsiveImage
        imageURL="https://i.pinimg.com/originals/a6/60/72/a66072b0e88258f2898a76c3f3c01041.jpg"
        alt="Who painted The Starry Night?"
      />,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render loading state', () => {
    const { container } = render(<ResponsiveImage alt="loading" />)
    expect(container).toMatchSnapshot()
  })

  it('should render error state', async () => {
    const { container } = render(
      <ResponsiveImage imageURL="invalid-url" alt="error" />,
    )
    // ensure effects (phase transition) have flushed
    await act(async () => {})
    expect(container).toMatchSnapshot()
  })

  it('should render with blur effect', () => {
    const { container } = render(
      <ResponsiveImage
        imageURL="https://i.pinimg.com/originals/a6/60/72/a66072b0e88258f2898a76c3f3c01041.jpg"
        alt="blur"
        revealEffect={{ type: QuestionImageRevealEffectType.Blur }}
      />,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render with square effect', () => {
    const { container } = render(
      <ResponsiveImage
        imageURL="https://i.pinimg.com/originals/a6/60/72/a66072b0e88258f2898a76c3f3c01041.jpg"
        alt="square"
        revealEffect={{ type: QuestionImageRevealEffectType.Square3x3 }}
      />,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render with overlay children', () => {
    const { container } = render(
      <ResponsiveImage
        imageURL="https://i.pinimg.com/originals/a6/60/72/a66072b0e88258f2898a76c3f3c01041.jpg"
        alt="overlay">
        <span>Overlay Content</span>
      </ResponsiveImage>,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render with noBorder', () => {
    const { container } = render(
      <ResponsiveImage
        imageURL="https://i.pinimg.com/originals/a6/60/72/a66072b0e88258f2898a76c3f3c01041.jpg"
        alt="noBorder"
        noBorder
      />,
    )
    expect(container).toMatchSnapshot()
  })

  it('computes box size from container and intrinsic dims (800x600 from 400x300 * 2)', async () => {
    const { container } = render(
      <ResponsiveImage imageURL="https://cdn/img.jpg" alt="computed" />,
    )

    // allow Image.onload & state updates to flush
    await act(async () => {})

    // wrapper that appears only in ready+box state
    const box = container.querySelector('[class*="box"]')
    expect(box).toBeTruthy()
    // inline styles
    expect((box as HTMLElement).style.width).toBe('800px')
    expect((box as HTMLElement).style.height).toBe('600px')

    const img = container.querySelector('img')
    expect(img).toHaveAttribute('src', 'https://cdn/img.jpg')
    expect(img).toHaveAttribute('alt', 'computed')
  })

  it('shows loading overlay while image is loading (before onload fires)', async () => {
    // Temporarily override Image to never resolve
    class LoadingImage {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      onload: null | ((ev?: any) => any) = null
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      onerror: null | ((ev?: any) => any) = null
      naturalWidth = 0
      naturalHeight = 0
      set src(_val: string) {
        // never call onload/onerror -> stays "loading"
      }
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const prev = (globalThis as any).Image
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ;(globalThis as any).Image = LoadingImage

    const { container, unmount } = render(
      <ResponsiveImage imageURL="https://cdn/slow.jpg" alt="loading" />,
    )
    // ensure effects (phase=loading) are applied
    await act(async () => {})

    expect(screen.getByTestId('spinner')).toBeInTheDocument()
    expect(container.querySelector('img')).toBeNull()

    unmount()
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ;(globalThis as any).Image = prev
  })

  it('shows error overlay when image fails to load', async () => {
    const { container } = render(<ResponsiveImage imageURL="bad" alt="err" />)
    await act(async () => {}) // allow onerror phase to apply
    // error overlay (Typography mocked)
    expect(screen.getByTestId('typography')).toHaveTextContent(
      'Oh no! Unable to load image',
    )
    // no img rendered
    const img = container.querySelector('img')
    expect(img).toBeNull()
  })

  it('applies Blur effect style via hook and does NOT render square effect', async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const countdown = { remainingMs: 3000 } as any
    const { container } = render(
      <ResponsiveImage
        imageURL="https://cdn/blur.jpg"
        alt="blur"
        revealEffect={{ type: QuestionImageRevealEffectType.Blur, countdown }}
      />,
    )
    await act(async () => {})

    const img = container.querySelector('img') as HTMLImageElement
    // Blur style returned from mocked hook
    expect(img.getAttribute('style') || '').toMatch(/blur\(8px\)/i)

    // No square effect component for Blur
    expect(screen.queryByTestId('square-effect')).toBeNull()
  })

  it('renders ImageSquareEffect for Square3x3 and passes box + effect', async () => {
    const { container } = render(
      <ResponsiveImage
        imageURL="https://cdn/tiles.jpg"
        alt="squares"
        revealEffect={{ type: QuestionImageRevealEffectType.Square3x3 }}
      />,
    )
    await act(async () => {})

    const square = screen.getByTestId('square-effect')
    expect(square).toBeInTheDocument()
    // From earlier computed box (800x600)
    expect(square).toHaveAttribute('data-w', '800')
    expect(square).toHaveAttribute('data-h', '600')
    expect(square).toHaveAttribute(
      'data-effect',
      String(QuestionImageRevealEffectType.Square3x3),
    )

    // img rendered normally (no blur style)
    const img = container.querySelector('img') as HTMLImageElement
    expect(img.getAttribute('style') || '').not.toMatch(/blur/i)
  })

  it('respects noBorder by using the boxNoBorder class', async () => {
    const { container } = render(
      <ResponsiveImage
        imageURL="https://cdn/noborder.jpg"
        alt="noBorder"
        noBorder
      />,
    )
    await act(async () => {})

    const box = container.querySelector('[class*="boxNoBorder"]')
    expect(box).toBeTruthy()
    const bordered = container.querySelector(
      '[class*="box"]:not([class*="boxNoBorder"])',
    )
    expect(bordered).toBeNull()
  })

  it('renders children into overlay when provided', async () => {
    const { container } = render(
      <ResponsiveImage imageURL="https://cdn/overlay.jpg" alt="overlay">
        <span data-testid="overlay-child">Overlay Content</span>
      </ResponsiveImage>,
    )
    await act(async () => {})

    const overlayChild = screen.getByTestId('overlay-child')
    expect(overlayChild).toBeInTheDocument()

    // ensure overlay wrapper exists
    const overlay = container.querySelector('[class*="overlay"]')
    expect(overlay).toBeTruthy()
  })

  it('resets to idle (no content) when imageURL becomes undefined after being set', async () => {
    const { rerender, container } = render(
      <ResponsiveImage imageURL="https://cdn/once.jpg" alt="once" />,
    )
    await act(async () => {})
    expect(container.querySelector('img')).toBeTruthy()

    // Drop URL -> should clear displaySrc/phase/intrinsic and render nothing (no spinner, no error)
    rerender(<ResponsiveImage alt="idle" />)
    await act(async () => {})

    expect(container.querySelector('img')).toBeNull()
    expect(screen.queryByTestId('spinner')).toBeNull()
    expect(screen.queryByTestId('typography')).toBeNull()
  })

  it('renders nothing in ready phase if container has no measurable size (box=null)', async () => {
    // For this test, report no size at all (for all observer calls during this render)
    resizeObserverImpl.mockImplementation(
      // eslint-disable-next-line @typescript-eslint/no-unused-vars,@typescript-eslint/no-explicit-any
      (_args: any) => {
        /* no onResize call => width/height remain 0 */
      },
    )

    const { container } = render(
      <ResponsiveImage imageURL="https://cdn/nosize.jpg" alt="nosize" />,
    )
    await act(async () => {})

    // phase becomes ready internally, but box=null => no img content
    expect(container.querySelector('img')).toBeNull()
    // Also not loading or error
    expect(screen.queryByTestId('spinner')).toBeNull()
    expect(screen.queryByTestId('typography')).toBeNull()

    // reset mock for subsequent tests
    resizeObserverImpl.mockReset()
  })
})
