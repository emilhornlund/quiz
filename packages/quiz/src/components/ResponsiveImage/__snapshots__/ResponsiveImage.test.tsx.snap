// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResponsiveImage > should render ResponsiveImage with props 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="centerOverlay"
    >
      <div
        class="loadingSpinnerContainer"
        data-testid="loading-spinner"
      >
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>
</div>
`;
